kernel/environ.ml

```diff
@@ -86,6 +86,7 @@ type env = {
   irr_inds : Sorts.relevance Indmap_env.t;
   symb_pats: rewrite_rule list Cmap_env.t;
   env_typing_flags  : typing_flags;
+  vm_library : Vmlibrary.t;
```

Comment:
This change seems a bit worrying to me. Are we sure making `env` mutable is a good idea?

That would render for example physical equality not faithful anymore. You could argue the "fetch" event is "benign", but actually things are way more complex than that, under-the-hood IO can make life of document managers pretty hard for example.

The need for efficient caching is a very common need to compiler/interpreters, (see for example Rust's use of Salsa), however I think we should refrain from adding N custom caches to Coq and instead discuss a common framework.

Reply:
While I agree that sneaky IO can be problematic, the VM / native state is already largely imperative, both inside and outside of the env type. Outside, we have a global mutable table of linked VM code, and similar hidden state through dynlink for native. Inside of the env type, we have actual mutable references wrapped in CEphemerons (more refs!) that cache the computation of the VM value, see Environ.key.

Reply:
Indeed, `Environ.key` is also problematic if the upper layers want to assume `env` is pure, adding even more imperative state and IO to envs is IMHO something we should consider carefully, and maybe even fix the current link hack to a proper solution that can be managed by a proper cache.

In particular the tradeoff with this PR is delicate, in the sense that while loading things under demand is highly desirable, the gains are not visible, and the impact on a core type like env is big (from the point of view of the invariants upper layers want)

For example, `coq-lsp` needs to assume `envs` are pure, otherwise it is just not possible to implement a system with the specs it has in a reasonable way.

Having good caching and memoization is a core need of a system like Coq, however experience shows we need to do a bit more principled than adding N random ad-hoc caches thru the codebase, and hacking vo files to serve as a poor man cache.